sycl/doc/extensions/proposed/sycl_ext_oneapi_device_wait.asciidoc

```diff
@@ -0,0 +1,175 @@
+
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+:endnote: &#8212;{nbsp}end{nbsp}note
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2025 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 10 specification.
+All references below to the "core SYCL specification" or to section numbers in
+the SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be in
+a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension adds a way for the host to wait for all commands submitted to a
+device to complete.
+This functionality is similar to the CUDA API `cudaDeviceSynchronize`.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_DEVICE_WAIT` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New aspect
+
+This extension adds the following aspect.
+
+[source,c++]
+----
+namespace sycl {
+
+enum class aspect {
+  // ...
+  ext_oneapi_device_wait
+};
+
+} // namespace sycl
+----
+
+'''
+
+[source,c++]
+----
+ext_oneapi_device_wait
+----
+
+Indicates that the device supports the `device::ext_oneapi_wait_and_throw`
+member functions.
+
+'''
+
+=== New member functions for the device class
+
+This extension adds the following member functions to the `device` class.
+
+[source,c++]
+----
+namespace sycl {
+
+class device {
+  // ...
+  void ext_oneapi_wait_and_throw(async_handler h);
+  void ext_oneapi_wait_and_throw();
```

Comment:
The semantic of these operations is close to CUDA but not exact.  I think the precise definition of `cudaDeviceSynchronize` is that it waits for all tasks submitted to the **current context** of the current device to complete.  (Credit to Jinghui for noticing this a couple weeks ago.)  Do we think it's important for migrating CUDA code to have this same semantic?

If so, I think we could:

* Add a new overload `ext_oneapi_wait_and_throw(const context& ctxt, async_handler h)`, which waits only for commands submitted to this device AND that context to complete.
* Redefine the other two overloads to mean that they wait only for commands submitted to this device AND its default context to complete.

This would require changing the proposed Level Zero API to also take a context.

Reply:
This doesn't really say anything, though, about how people use the CUDA API or what semantics they would want in SYCL in order to migrate their code to SYCL.  I think you are just pointing out that the currently proposed API would be easy to implement on our current design of the UR CUDA adaptor.